### PR TITLE
Unifying instantiate2 implementation

### LINKED ISSUE #207: Unifying `instantiate2` implementation
## Description

Our team at @AbstractSDK is working on refactoring our smart-account implementations to be optimized for account-abstraction use-cases. Because each user on the blockchain will have an Abstract smart-account created for them, we're trying to minimize the state that is stored under the contract's namespace. 

The account needs to be aware of a few contracts in order to do verification of actions or call specific functionality on them. 

## Our Solution

To reduce the state we've been exploring the use of `instantiate2` to deterministically compute the canonical address of each contract and include them in the WASM binary directly as bytes, to be humanized at runtime. As the # of Accounts scale, this should have a significant impact on the storage use of our solution. 

To get the address of any of these native contracts we do the following: 
- Upload a CosmWasm "blob" contract that is a minimal CW contract.
- Use `instantiate2` with a contract-specific salt to claim the address for that specific contract. 
- Migrate the blob on that address to the contract that we want to have occupy that address.

## The Problem

The problem that we're facing is that this workflow does not give us the same addresses in the `cw-multi-test` environment due to the fact that its `instantiate2` method does not incorporate the actual checksum of the blob contract. 

As a result we're forced to either: 
- Construct our own `ChecksumGenerator` and use it in all our `cw-multi-test` uses. 
  - We don't like this solution because as a development framework, we'd require all our users to use this exact `ChecksumGenerator`, which might conflict with their own requirements.
- Use feature-flags to select a set of different addresses for use in cw-multi-test.
  - This will create conflicts in our codebase, requiring us to separate cw-multi-test tests from all other kinds of tests and would create a lot of complexity in our codebase overall. 
  
## Discussed Solution

We've internally discussed an addition to `cw-multi-test` that would allow users to set the `checksum` of a contract through the `Contract` trait.

```rust
pub trait Contract<C, Q = Empty>
where
    C: CustomMsg,
    Q: CustomQuery,
{
    /// Get the checksum of the contract
    // Defaults to None (not API breaking)
    fn checksum(&self) -> AnyResult<Option<String>> {
        Ok(None)
   }
}
```
And then update `ContractWrapper` to contain an optional `checksum` value.
```rust
pub struct ContractWrapper<
   ...
> where
    ...
{
    // Add an optional checksum field
    checksum: Option<String>,
}
```

And update the `instantiate2` method to use this checksum if `Some`. 

Doing this would allow developers to use the exact same addresses for their contracts in a test environment as a `wasmd` environment without requiring custom checksum generators.

## Questions

Is this an acceptable solution?

### PR DESCRIPTION
**Pull request by DariuszDepta**
Added `checksum()` function to `Contract` trait and `checksum` field to `ContractWrapper`. This checksum (when provided) is used to generate the predictable contract address (when processing the instantiate2 message).

closes #207

### CODE REVIEW

**Review by Buckram123 — APPROVED** [review]

File: tests/test_app/test_instantiate2.rs [hunk 1]
```diff
@@ -38,34 +44,48 @@ fn instantiate2_works() {
...
-    // humanize the address of the contract
-    let contract_human_addr = app.api().addr_humanize(&contract_addr).unwrap();
+    // check the instantiate result
+    let parsed = parse_instantiate_response_data(res.data.unwrap().as_slice()).unwrap();
+    assert!(parsed.data.is_none());
 
-    // check if the predictable contract's address matches the result from instantiate2_address function
-    assert_eq!(parsed.contract_address, contract_human_addr.to_string());
+    // check the resulting predictable contract's address
+    assert_eq!(parsed.contract_address, JUNO_2);
```
[the remark this answers is not in the document]
Comment (CyberHoward) [reply tests/test_app/test_instantiate2.rs:85]: Nevermind, it's performed in `additional_check`!

**Review by CyberHoward — APPROVED** [review]

File: src/contracts.rs [hunk 2]
```diff
@@ -33,6 +33,11 @@ where
 
     /// Evaluates contract's `migrate` entry-point.
     fn migrate(&self, deps: DepsMut<Q>, env: Env, msg: Vec<u8>) -> AnyResult<Response<C>>;
+
+    /// Returns the _simulated_ checksum of the contract's Wasm blob.
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/contracts.rs]: Good catch, changed it to `provided`. Makes more sense.

File: tests/test_app/test_instantiate2.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply tests/test_app/test_instantiate2.rs]: I have added comments in these tests to make it more clear that the comparison with the addresses generated by CW VM are done in function `compare_with_cosmwasm_vm_address` (renamed `additional_check` => `compare_with_cosmwasm_vm_address` to make it more clear).

**Comment by DariuszDepta** [comment]
@Buckram123 @CyberHoward thanks a lot for your review!

**Pull request merged**